### PR TITLE
sd-dhcp6-client: several fixes and cleanups

### PR DESCRIPTION
**Pull request by yuwata**

### CODE REVIEW

**Comment by yuwata** [comment]
cc @evverx and @thom311

**Comment by evverx** [comment]
@yuwata thanks! I'm not sure `assert_return` should be converted into `assert` unconditionally because unfortunately this code isn't covered by enough tests to be absolutely sure that rogue packets aren't going to crash it. I think it should be configurable with maybe a meson option like `-Dabort-assert-return` that could be passed on the CI but judging by https://github.com/systemd/systemd/pull/22413 the CI isn't ready for that either.

**Comment by yuwata** [comment]
The converted functions are not exposed, hence, I hope they should not cause any real issues. And, if one of them is triggered, it is a critical error in sd-dhcp6-client, and we need to address it. I think this should be fine to merge this.

**Comment by thom311** [comment]
@yuwata I assume it that commit "sd-dhcp6-client: introduce two helpers to create message" replaces #22406. Is "sd-dhcp6-client: fix typo ia_pd -> ia_na" also strictly necessary?

The more important question is, whether the patch from #22406 sufficiently fixes the problem, or whether I need to port #22412 to NetworkManager ASAP.

**Comment by yuwata** [comment]
@thom311 After re-considering this and your PR, now I think we should merge yours at first, then this can be mostly cleanups and hopefully not necessary to be backported.

**Comment by yuwata** [comment]
Rebased on top of #22406. PTAL.

@thom311 I noticed the first commit should be also backported, though the issue fixed by the commit does not trigger assertion unlike the issue fixed by you.

**Comment by yuwata** [comment]
This becomes huge...

**Comment by evverx** [comment]
I didn't expect it to snowball into this either :-)

With asserts enabled the fuzzer fails with
```
INFO:     2678 files found in ./SERVER/
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 1048576 bytes
INFO: seed corpus: files: 2678 min: 1b max: 4194303b total: 10957530b rss: 59Mb
Code should not be reached at src/libsystemd-network/sd-dhcp6-client.c:738, function DHCP6MessageType client_message_type_from_state(sd_dhcp6_client *)(). Aborting. 💥
==6821== ERROR: libFuzzer: deadly signal
    #0 0x51b23e in __sanitizer_print_stack_trace (/home/vagrant/systemd/out/fuzz-dhcp6-client-send+0x51b23e)
    #1 0x4688f9 in fuzzer::PrintStackTrace() (/home/vagrant/systemd/out/fuzz-dhcp6-client-send+0x4688f9)
    #2 0x449d56 in fuzzer::Fuzzer::CrashCallback() (.part.0) FuzzerLoop.cpp.o
    #3 0x449e1b in fuzzer::Fuzzer::StaticCrashSignalCallback() (/home/vagrant/systemd/out/fuzz-dhcp6-client-send+0x449e1b)
    #4 0x7f3114f2174f  (/lib64/libc.so.6+0x4274f)
    #5 0x7f3114f6e84b in __pthread_kill_implementation (/lib64/libc.so.6+0x8f84b)
    #6 0x7f3114f216a5 in gsignal (/lib64/libc.so.6+0x426a5)
    #7 0x7f3114f0b7d2 in abort (/lib64/libc.so.6+0x2c7d2)
    #8 0x7f3115f4499d in log_assert_failed_unreachable /home/vagrant/systemd/build/../src/basic/log.c:875:9
    #9 0x55b0f2 in client_message_type_from_state /home/vagrant/systemd/build/../src/libsystemd-network/sd-dhcp6-client.c:738:17
    #10 0x55b0f2 in client_send_message /home/vagrant/systemd/build/../src/libsystemd-network/sd-dhcp6-client.c:769:25
    #11 0x55826c in LLVMFuzzerTestOneInput /home/vagrant/systemd/build/../src/libsystemd-network/fuzz-dhcp6-client-send.c:50:16
    #12 0x44a289 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/home/vagrant/systemd/out/fuzz-dhcp6-client-send+0x44a289)
    #13 0x44adf0 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) (/home/vagrant/systemd/out/fuzz-dhcp6-client-send+0x44adf0)
    #14 0x44d4b9 in fuzzer::Fuzzer::ReadAndExecuteSeedCorpora(std::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) (/home/vagrant/systemd/out/fuzz-dhcp6-client-send+0x44d4b9)
    #15 0x44d970 in fuzzer::Fuzzer::Loop(std::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) (/home/vagrant/systemd/out/fuzz-dhcp6-client-send+0x44d970)
    #16 0x4328da in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/home/vagrant/systemd/out/fuzz-dhcp6-client-send+0x4328da)
    #17 0x421e96 in main (/home/vagrant/systemd/out/fuzz-dhcp6-client-send+0x421e96)
    #18 0x7f3114f0c55f in __libc_start_call_main (/lib64/libc.so.6+0x2d55f)
    #19 0x7f3114f0c60b in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x2d60b)
    #20 0x421ee4 in _start (/home/vagrant/systemd/out/fuzz-dhcp6-client-send+0x421ee4)

NOTE: libFuzzer has rudimentary signal handlers.
      Combine libFuzzer with AddressSanitizer or similar for better crash reports.
SUMMARY: libFuzzer: deadly signal
```
I'm not sure whether it's a dhcp-client issue. Looks like the fuzz target needs tweaking.

**Comment by evverx** [comment]
Judging by
```
$1 = {n_ref = 1, state = DHCP6_STATE_BOUND, event = 0x617000000080, event_priority = 0, ifindex = 42, ifname = 0x0, local_address = {__in6_u = {
      __u6_addr8 = "\376\200", '\000' <repeats 13 times>, "\001", __u6_addr16 = {33022, 0, 0, 0, 0, 0, 0, 256}, __u6_addr32 = {33022, 0, 0, 16777216}}}, hw_addr = {length = 0, {ether = {
        ether_addr_octet = "\000\000\000\000\000"}, infiniband = '\000' <repeats 19 times>, in = {s_addr = 0}, in6 = {__in6_u = {__u6_addr8 = '\000' <repeats 15 times>, __u6_addr16 = {0, 0, 0, 0, 0, 0,
            0, 0}, __u6_addr32 = {0, 0, 0, 0}}}, bytes = '\000' <repeats 31 times>}}, arp_type = 0, ia_na = {type = 3, header = {id = 1243341913, lifetime_t1 = 0, lifetime_t2 = 0}, addresses = 0x0},
  ia_pd = {type = 25, header = {id = 1243341913, lifetime_t1 = 0, lifetime_t2 = 0}, addresses = 0x0}, request_ia = (DHCP6_REQUEST_IA_NA | DHCP6_REQUEST_IA_PD), transaction_id = 1685289216,
  transaction_start = 17007157783, lease = 0x611000000400, fd = 4, information_request = false, iaid_set = true, req_opts = 0x6020000001d0, req_opts_len = 4, fqdn = 0x6020000001f0 "example.com",
  mudurl = 0x604000000190 "https://www.example.com/mudfile.json", user_class = 0x603000000460, vendor_class = 0x603000000490, receive_message = 0x612000000340, retransmit_time = 0,
  retransmit_count = 0 '\000', timeout_resend = 0x6120000004c0, timeout_expire = 0x612000000940, timeout_t1 = 0x612000000640, timeout_t2 = 0x6120000007c0, callback = 0x0, userdata = 0x0, duid = {
    type = 512, {llt = {htype = 0, time = 2002850219, haddr = 0x615000000930 "@\336\023Bâ"}, en = {pen = 296419328, id = "aw@\336\023Bâ"}, ll = {htype = 0,
        haddr = 0x61500000092c "\253\021aw@\336\023Bâ"}, uuid = {uuid = {bytes = "\000\000\253\021aw@\336\023Bâ\000\000\000", qwords = {16014931533721436160, 2730705427}}}, raw = {
        data = "\000\000\253\021aw@\336\023Bâ", '\000' <repeats 115 times>}}}, duid_len = 14, information_request_time_usec = 0, information_refresh_time_usec = 86400000000, extra_options = 0x0,
  vendor_options = 0x0, test_mode = true}
```
the state is `DHCP6_STATE_BOUND` when it crashes so it seems the fist client state is no longer `DHCP6_STATE_SOLICITATION`

**Comment by yuwata** [comment]
@evverx The issue you found should be now fixed. PTAL.

**Comment by evverx** [comment]
@yuwata thanks! Looks like it's gone. But looking at `client_process_advertise_or_rapid_commit_reply` in that fuzzer I noticed that packets seem to be parsed using 3 functions, which makes `fuzz-dhcp6-client` kind of out of date. It triggers only `client_receive_reply` and `client_process_advertise_or_rapid_commit_reply` as far as I can tell and `client_process_information` isn't covered by any fuzzers. Could you update it too?

Also looking at the commit where vendor options were fixed it seems I forgot to add them to the fuzzer along with `vendor_class`, `user_class` and so on. Could you add vendor options to `fuzz-dhcp6-client-send` to let it cover that memory leak? It would probably make sense to update `test-networkd.py` as well. It doesn't cover that at all as far as I can see.

**Comment by evverx** [comment]
FWIW personally I'm kind of hesitant to review large PRs like this. If it was possible to split them into a few PRs where the tests could be updated along with the code as well it would be ideal. It's probably too late here but I think in general PRs should be more reviewable so to speak :-) @mrc0mmand it would be great if you could take a look and help to figure out whether `test-networkd.py` needs tweaking. Thanks!

**Comment by yuwata** [comment]
@evverx Sure. Unfortunately, even this is already huge, there are many pending commits, and I am also planning to update/improve tests and fuzzers.

**Comment by mrc0mmand** [comment]
> FWIW personally I'm kind of hesitant to review large PRs like this. If it was possible to split them into a few PRs where the tests could be updated along with the code as well it would be ideal. It's probably too late here but I think in general PRs should be more reviewable so to speak :-) @mrc0mmand it would be great if you could take a look and help to figure out whether `test-networkd.py` needs tweaking. Thanks!

I guess I should look into running `test-networkd.py` alongside `systemd-networkd-testsuite.py` in CentOS CI again. Since am on an unexpected PTO today I'll give it a shot tomorrow.

**Comment by evverx** [comment]
> The fuzzer is improved

CIFuzz doesn't report timeout by default but It seems the fuzzer just times out
```


artifact_prefix='./'; Test unit written to ./timeout-da39a3ee5e6b4b0d3255bfef95601890afd80709
Base64:
==8804== ERROR: libFuzzer: timeout after 31 seconds
    #0 0x51b26e in __sanitizer_print_stack_trace (/home/vagrant/systemd/out/fuzz-dhcp6-client+0x51b26e)
    #1 0x468929 in fuzzer::PrintStackTrace() (/home/vagrant/systemd/out/fuzz-dhcp6-client+0x468929)
    #2 0x44a034 in fuzzer::Fuzzer::StaticAlarmCallback() (/home/vagrant/systemd/out/fuzz-dhcp6-client+0x44a034)
    #3 0x7f1ce364174f  (/lib64/libc.so.6+0x4274f)
    #4 0x7f1ce3710c1d in epoll_wait (/lib64/libc.so.6+0x111c1d)
    #5 0x7f1ce4a67cec in epoll_wait_usec /home/vagrant/systemd/build/../src/libsystemd/sd-event/sd-event.c:3946:27
    #6 0x7f1ce4a67cec in process_epoll /home/vagrant/systemd/build/../src/libsystemd/sd-event/sd-event.c:3969:21
    #7 0x7f1ce4a67cec in sd_event_wait /home/vagrant/systemd/build/../src/libsystemd/sd-event/sd-event.c:4088:21
    #8 0x7f1ce4a763b1 in sd_event_run /home/vagrant/systemd/build/../src/libsystemd/sd-event/sd-event.c:4237:21
    #9 0x54b6a9 in fuzz_client /home/vagrant/systemd/build/../src/libsystemd-network/fuzz-dhcp6-client.c:46:9
    #10 0x54a9c0 in LLVMFuzzerTestOneInput /home/vagrant/systemd/build/../src/libsystemd-network/fuzz-dhcp6-client.c:109:9
    #11 0x44a2b9 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/home/vagrant/systemd/out/fuzz-dhcp6-client+0x44a2b9)
    #12 0x44d435 in fuzzer::Fuzzer::ReadAndExecuteSeedCorpora(std::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) (/home/vagrant/systemd/out/fuzz-dhcp6-client+0x44d435)
    #13 0x44d9a0 in fuzzer::Fuzzer::Loop(std::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) (/home/vagrant/systemd/out/fuzz-dhcp6-client+0x44d9a0)
    #14 0x43290a in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/home/vagrant/systemd/out/fuzz-dhcp6-client+0x43290a)
    #15 0x421ec6 in main (/home/vagrant/systemd/out/fuzz-dhcp6-client+0x421ec6)
    #16 0x7f1ce362c55f in __libc_start_call_main (/lib64/libc.so.6+0x2d55f)
    #17 0x7f1ce362c60b in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x2d60b)
    #18 0x421f14 in _start (/home/vagrant/systemd/out/fuzz-dhcp6-client+0x421f14)

SUMMARY: libFuzzer: timeout
```

**Comment by evverx** [comment]
CIFuzz actually reported it but looking at https://github.com/google/oss-fuzz/pull/6997 and https://github.com/google/oss-fuzz/pull/6711 I think it's kind of a bug :-)

**Comment by yuwata** [comment]
@evverx Is it possible to get the reproducer of the failed fuzzers?

**Comment by evverx** [comment]
I think it can be reproduced with
```
./tools/oss-fuzz.sh
./out/fuzz-dhcp6-client -timeout=30 test/fuzz/fuzz-dhcp6-client
```

**Comment by yuwata** [comment]
@evverx Thank you. I hope I found the root cause.

File: src/libsystemd-network/dhcp6-option.c [hunk 1]
```diff
@@ -317,7 +317,7 @@ int dhcp6_option_append_ia(uint8_t **buf, size_t *buflen, const DHCP6IA *ia) {
                 break;
 
         default:
-                return -EINVAL;
+                assert_not_reached();
```
Comment (bluca) [inline src/libsystemd-network/dhcp6-option.c]: commit message says `refuse invalid IA type` but this is more than refuse, it's an assert - commit message should be updated to reflect it

Comment (bluca) [reply src/libsystemd-network/dhcp6-option.c]: (commit 7a4f4b9a3c772668b283f75fd2de1b9660f3f2bf "sd-dhcp6-client: make dhcp6_option_append_ia() refuse invalid IA type" )

File: src/libsystemd-network/sd-dhcp6-client.c [hunk 2]
```diff
@@ -863,34 +863,36 @@ static int client_enter_bound_state(sd_dhcp6_client *client) {
...
+        }
 
-        r = event_reset_time_relative(client->event, &client->timeout_t2,
-                                      clock_boottime_or_monotonic(),
-                                      timeout, 10 * USEC_PER_SEC,
-                                      client_timeout_t2, client,
-                                      client->event_priority, "dhcp6-t2-timeout", true);
-        if (r < 0)
-                goto error;
+        if (lifetime_t1 == USEC_INFINITY) {
```
Comment (bluca) [inline src/libsystemd-network/sd-dhcp6-client.c]: did you mean `lifetime_t2 == USEC_INFINITY` here?

File: src/libsystemd-network/sd-dhcp6-client.c [hunk 3]
```diff
@@ -951,8 +951,6 @@ static int client_timeout_resend(sd_event_source *s, uint64_t usec, void *userda
         assert(s);
         assert(client->event);
 
-        (void) event_source_disable(client->timeout_resend);
```
Comment (bluca) [inline src/libsystemd-network/sd-dhcp6-client.c:981]: This function can return before re-enabling this, won't it be a problem?

File: src/libsystemd-network/sd-dhcp6-client.c [hunk 4]
```diff
@@ -1388,7 +1376,8 @@ int sd_dhcp6_client_start(sd_dhcp6_client *client) {
         if (!client->information_request && client->request_ia == 0)
                 return -EINVAL;
 
-        client_reset(client);
+        /* Even the client in STOPPED state, the lease acquired in the previous information request may stored. */
```
Comment (bluca) [inline src/libsystemd-network/sd-dhcp6-client.c]: "Even if the client is in the STOPPED state"
"may be stored"

File: src/libsystemd-network/sd-dhcp6-client.c [hunk 5]
```diff
@@ -191,7 +191,9 @@ int sd_dhcp6_client_set_mac(
         assert_return(client, -EINVAL);
         assert_return(addr, -EINVAL);
         assert_return(addr_len <= HW_ADDR_MAX_SIZE, -EINVAL);
-        assert_return(client->state == DHCP6_STATE_STOPPED, -EBUSY);
+
+        /* Unlike the other setters, it is OK to set new MAC address during the client is running,
```
Comment (bluca) [inline src/libsystemd-network/sd-dhcp6-client.c]: "it is OK to set a new MAC address while the client is running,"

File: src/libsystemd-network/dhcp-identifier.c [hunk 6]
```diff
@@ -120,20 +120,23 @@ static int dhcp_identifier_set_duid_ll(const uint8_t *addr, size_t addr_len, uin
...
-                return r;
-#else
-        machine_id = SD_ID128_MAKE(01, 02, 03, 04, 05, 06, 07, 08, 09, 0a, 0b, 0c, 0d, 0e, 0f, 10);
-#endif
+        if (!test_mode) {
+                r = sd_id128_get_machine(&machine_id);
+                if (r < 0)
+                        return r;
+        } else
+                /* For tests, especially for fuzzers, it is important the reproducibility.
```
Comment (bluca) [inline src/libsystemd-network/dhcp-identifier.c]: "reproducibility is important"

File: src/libsystemd-network/dhcp-identifier.c [hunk 7]
```diff
@@ -120,20 +120,23 @@ static int dhcp_identifier_set_duid_ll(const uint8_t *addr, size_t addr_len, uin
...
+                 * Hence, use the static and constant machine id.
```
Comment (bluca) [inline src/libsystemd-network/dhcp-identifier.c]: "use a static"

File: src/libsystemd-network/sd-dhcp6-client.c [hunk 8]
```diff
@@ -631,6 +631,11 @@ static void client_stop(sd_dhcp6_client *client, int error) {
 
         client->lease = sd_dhcp6_lease_unref(client->lease);
 
+        /* Reset IRT here. Otherwise, we cannot restart the client in information requesting mode,
+         * even the lease is freed below. */
```
Comment (bluca) [inline src/libsystemd-network/sd-dhcp6-client.c]: "even though the lease is freed below"

**Review by bluca — APPROVED** [review]
That was a lot of commits! Still, the last one is pretty large, maybe split into at least two, with the new rapid commit test separately? And a few more nits

File: src/libsystemd-network/sd-dhcp6-client.c (hunk 3 above)
Comment (yuwata) [reply src/libsystemd-network/sd-dhcp6-client.c:981]: When it returns earlier, then it calls `client_set_state()` or `client_stop()`, and both disable the timer. Moreover, timer event source is one shot by default, so it is not necessary to disable it in the callback.

Comment (yuwata) [reply src/libsystemd-network/sd-dhcp6-client.c:981]: The commit message now mention about that.

**Comment by yuwata** [comment]
@bluca Thank you for reviewing such a huge PR. The above points are addressed. Upgrading the green label.

File: src/libsystemd-network/dhcp6-lease-internal.h [hunk 9]
```diff
@@ -23,8 +23,8 @@ struct sd_dhcp6_lease {
         triple_timestamp timestamp;
         struct in6_addr server_address;
 
-        DHCP6IA ia;
-        DHCP6IA pd;
+        DHCP6IA ia_na;
```
Comment (poettering) [inline src/libsystemd-network/dhcp6-lease-internal.h]: maybe even add a comment what cryptic "IA NA" means…

File: src/libsystemd-network/dhcp-identifier.c [hunk 10]
```diff
@@ -57,110 +64,146 @@ int dhcp_validate_duid_len(uint16_t duid_type, size_t duid_len, bool strict) {
...
         if (r < 0)
                 return r;
 
-        unaligned_write_be16(&duid->type, DUID_TYPE_UUID);
-        memcpy(&duid->raw.data, &machine_id, sizeof(machine_id));
+        unaligned_write_be16(&ret_duid->type, DUID_TYPE_UUID);
+        memcpy(ret_duid->raw.data, &machine_id, sizeof(machine_id));
 
-        *len = sizeof(duid->type) + sizeof(machine_id);
+        *ret_len = sizeof(ret_duid->type) + sizeof(machine_id);
```
Comment (poettering) [inline src/libsystemd-network/dhcp-identifier.c:171]: maybe:

```c
*ret_len = offsetof(struct duid, raw.data) + sizeof(machine_id);
```

File: src/libsystemd-network/dhcp-identifier.c [hunk 11]
```diff
@@ -57,110 +64,146 @@ int dhcp_validate_duid_len(uint16_t duid_type, size_t duid_len, bool strict) {
...
-        unaligned_write_be16(&duid->llt.htype, arp_type);
-        unaligned_write_be32(&duid->llt.time, time_from_2000y);
-        memcpy(duid->llt.haddr, addr, addr_len);
+        unaligned_write_be16(&ret_duid->type, DUID_TYPE_LLT);
+        unaligned_write_be16(&ret_duid->llt.htype, arp_type);
+        unaligned_write_be32(&ret_duid->llt.time, time_from_2000y);
+        memcpy(ret_duid->llt.haddr, addr, addr_len);
 
-        *len = sizeof(duid->type) + sizeof(duid->llt.htype) + sizeof(duid->llt.time) + addr_len;
+        *ret_len = sizeof(ret_duid->type) + sizeof(ret_duid->llt.htype) + sizeof(ret_duid->llt.time) + addr_len;
```
Comment (poettering) [inline src/libsystemd-network/dhcp-identifier.c:94]:
```c
*ret_len = offsetof(struct duid, llt.haddr) + addr_len;
```

File: src/libsystemd-network/dhcp-identifier.c [hunk 12]
```diff
@@ -57,110 +64,146 @@ int dhcp_validate_duid_len(uint16_t duid_type, size_t duid_len, bool strict) {
...
 
-        unaligned_write_be16(&duid->type, DUID_TYPE_LL);
-        unaligned_write_be16(&duid->ll.htype, arp_type);
-        memcpy(duid->ll.haddr, addr, addr_len);
+        unaligned_write_be16(&ret_duid->type, DUID_TYPE_LL);
+        unaligned_write_be16(&ret_duid->ll.htype, arp_type);
+        memcpy(ret_duid->ll.haddr, addr, addr_len);
 
-        *len = sizeof(duid->type) + sizeof(duid->ll.htype) + addr_len;
+        *ret_len = sizeof(ret_duid->type) + sizeof(ret_duid->ll.htype) + addr_len;
```
Comment (poettering) [inline src/libsystemd-network/dhcp-identifier.c:118]: similar

File: src/libsystemd-network/dhcp-identifier.c [hunk 13]
```diff
@@ -57,110 +64,146 @@ int dhcp_validate_duid_len(uint16_t duid_type, size_t duid_len, bool strict) {
...
+        unaligned_write_be16(&ret_duid->type, DUID_TYPE_EN);
+        unaligned_write_be32(&ret_duid->en.pen, SYSTEMD_PEN);
 
         /* a bit of snake-oil perhaps, but no need to expose the machine-id
          * directly; duid->en.id might not be aligned, so we need to copy */
         hash = htole64(siphash24(&machine_id, sizeof(machine_id), HASH_KEY.bytes));
-        memcpy(duid->en.id, &hash, sizeof(duid->en.id));
+        memcpy(ret_duid->en.id, &hash, sizeof(ret_duid->en.id));
+
+        *ret_len = sizeof(ret_duid->type) + sizeof(ret_duid->en);
```
Comment (poettering) [inline src/libsystemd-network/dhcp-identifier.c:149]: ditto

File: src/libsystemd-network/dhcp6-option.c [hunk 14]
```diff
@@ -385,57 +332,67 @@ static int option_append_pd_prefix(uint8_t **buf, size_t *buflen, const DHCP6Add
...
+        case SD_DHCP6_OPTION_IA_NA:
+        case SD_DHCP6_OPTION_IA_PD:
+                len = sizeof(struct ia_header);
+                header = (struct ia_header) {
+                        .id = ia->header.id,
+                };
+                break;
+
+        case SD_DHCP6_OPTION_IA_TA:
+                len = sizeof(be32_t); /* IA_TA does not have lifetime. */
```
Comment (poettering) [inline src/libsystemd-network/dhcp6-option.c:360]: maybe:

```c
len = sizeof(*header->id);
```

File: src/libsystemd-network/dhcp6-option.c [hunk 15]
```diff
@@ -385,57 +332,67 @@ static int option_append_pd_prefix(uint8_t **buf, size_t *buflen, const DHCP6Add
...
-        LIST_FOREACH(addresses, prefix, pd->addresses) {
-                r = option_append_pd_prefix(buf, buflen, prefix);
-                if (r < 0)
-                        return r;
-
-                len += r;
-        }
+        memcpy(*buf, &header, len);
+        *buf += len;
+        *buflen -= len;
```
Comment (poettering) [inline src/libsystemd-network/dhcp6-option.c:382]: maybe:

```c
*buf = mempcpy(*buf, &header, len);
buflen -= len;
``

**Comment by poettering** [comment]
lgtm. found some superficial things that could also be fixed. but they don't really matter. maybe in a different pr, given this one is large enough already.

**Comment by yuwata** [comment]
@poettering Thanks. I will address them after this is merged.

**Comment by evverx** [comment]
@yuwata Valgrind seems to report a few fd leaks:
```
==151950==
==151950== FILE DESCRIPTORS: 7 open (3 std) at exit.
==151950== Open AF_UNIX socket 5: <unknown>
==151950==    at 0x4DBB52E: socketpair (syscall-template.S:120)
==151950==    by 0x4063B2: dhcp6_network_bind_udp_socket (test-dhcp6-client.c:971)
==151950==    by 0x4122E2: sd_dhcp6_client_start (sd-dhcp6-client.c:1305)
==151950==    by 0x405D8B: test_client_callback (test-dhcp6-client.c:892)
==151950==    by 0x40DCAA: client_notify (sd-dhcp6-client.c:538)
==151950==    by 0x410251: client_enter_bound_state (sd-dhcp6-client.c:1012)
==151950==    by 0x410D18: client_process_reply (sd-dhcp6-client.c:1094)
==151950==    by 0x411EA2: client_receive_message (sd-dhcp6-client.c:1250)
==151950==    by 0x4B9CF80: source_dispatch (sd-event.c:3598)
==151950==    by 0x4B9F009: sd_event_dispatch (sd-event.c:4183)
==151950==    by 0x4B9F4EB: sd_event_run (sd-event.c:4244)
==151950==    by 0x4B9F6CF: sd_event_loop (sd-event.c:4265)
==151950==
==151950== Open AF_UNIX socket 6: <unknown>
==151950==    at 0x4DBB52E: socketpair (syscall-template.S:120)
==151950==    by 0x4063B2: dhcp6_network_bind_udp_socket (test-dhcp6-client.c:971)
==151950==    by 0x4122E2: sd_dhcp6_client_start (sd-dhcp6-client.c:1305)
==151950==    by 0x4066FC: test_dhcp6_client (test-dhcp6-client.c:997)
==151950==    by 0x406870: main (test-dhcp6-client.c:1018)
==151950==
==151950== Open file descriptor 4:
==151950==    at 0x4DBABFB: timerfd_create (syscall-template.S:120)
==151950==    by 0x4B92FEA: event_setup_timer_fd (sd-event.c:1152)
==151950==    by 0x4B93175: setup_clock_data (sd-event.c:1182)
==151950==    by 0x4B93649: sd_event_add_time (sd-event.c:1255)
==151950==    by 0x4B9381F: sd_event_add_time_relative (sd-event.c:1303)
==151950==    by 0x4064E8: test_dhcp6_client (test-dhcp6-client.c:982)
==151950==    by 0x406870: main (test-dhcp6-client.c:1018)
==151950==
==151950== Open file descriptor 3:
==151950==    at 0x4DBA89B: epoll_create1 (syscall-template.S:120)
==151950==    by 0x4B90251: sd_event_new (sd-event.c:399)
==151950==    by 0x406495: test_dhcp6_client (test-dhcp6-client.c:981)
==151950==    by 0x406870: main (test-dhcp6-client.c:1018)
```
I'm not sure whether it's a test issue or a client issue though. My guess would be that `sd_dhcp6_client_stop` is missing somewhere.

**Comment by yuwata** [comment]
@evverx I cannot reproduce the issue. `sd_dhcp6_client` is `unref()`ed by static destructor.
Do you run the test without `sudo`?

**Comment by yuwata** [comment]
> I'm not sure whether it's a test issue or a client issue though. My guess would be that `sd_dhcp6_client_stop` is missing somewhere.

`sd_dhcp6_client_unref()` should work without `sd_dhcp6_client_stop()`.

**Comment by yuwata** [comment]
BTW, please open a new issue if you find anything. Otherwise, we can easily forget your comment.

**Comment by evverx** [comment]
> BTW, please open a new issue if you find anything

I opened https://github.com/systemd/systemd/issues/22576 where I tried to collect all the fd leaks at once :-)

>  Do you run the test without sudo?

I ran it with `sudo` in https://github.com/evverx/systemd/pull/91 and there it reported
```
2022-02-21T02:21:34.9672293Z ==24413== FILE DESCRIPTORS: 7 open at exit.
2022-02-21T02:21:34.9672563Z ==24413== Open AF_UNIX socket 5: <unknown>
2022-02-21T02:21:34.9672979Z ==24413==    at 0x4DC293E: socketpair (syscall-template.S:78)
2022-02-21T02:21:34.9673455Z ==24413==    by 0x10FF41: dhcp6_network_bind_udp_socket (test-dhcp6-client.c:971)
2022-02-21T02:21:34.9673953Z ==24413==    by 0x11B961: sd_dhcp6_client_start (sd-dhcp6-client.c:1305)
2022-02-21T02:21:34.9674423Z ==24413==    by 0x10F872: test_client_callback (test-dhcp6-client.c:892)
2022-02-21T02:21:34.9674864Z ==24413==    by 0x1174A7: client_notify (sd-dhcp6-client.c:538)
2022-02-21T02:21:34.9675323Z ==24413==    by 0x119984: client_enter_bound_state (sd-dhcp6-client.c:1012)
2022-02-21T02:21:34.9675880Z ==24413==    by 0x11A401: client_process_reply (sd-dhcp6-client.c:1094)
2022-02-21T02:21:34.9676381Z ==24413==    by 0x11B55A: client_receive_message (sd-dhcp6-client.c:1250)
2022-02-21T02:21:34.9676836Z ==24413==    by 0x4B905CC: source_dispatch (sd-event.c:3598)
2022-02-21T02:21:34.9677245Z ==24413==    by 0x4B92508: sd_event_dispatch (sd-event.c:4183)
2022-02-21T02:21:34.9678236Z ==24413==    by 0x4B929C4: sd_event_run (sd-event.c:4244)
2022-02-21T02:21:34.9678776Z ==24413==    by 0x4B92B89: sd_event_loop (sd-event.c:4265)
2022-02-21T02:21:34.9680519Z ==24413== 
2022-02-21T02:21:34.9681040Z ==24413== Open AF_UNIX socket 6: <unknown>
2022-02-21T02:21:34.9682172Z ==24413==    at 0x4DC293E: socketpair (syscall-template.S:78)
2022-02-21T02:21:34.9685844Z ==24413==    by 0x10FF41: dhcp6_network_bind_udp_socket (test-dhcp6-client.c:971)
2022-02-21T02:21:34.9686348Z ==24413==    by 0x11B961: sd_dhcp6_client_start (sd-dhcp6-client.c:1305)
2022-02-21T02:21:34.9686782Z ==24413==    by 0x1102E2: test_dhcp6_client (test-dhcp6-client.c:997)
2022-02-21T02:21:34.9687212Z ==24413==    by 0x11047A: main (test-dhcp6-client.c:1018)
2022-02-21T02:21:34.9687818Z ==24413== 
2022-02-21T02:21:34.9688458Z ==24413== Open file descriptor 4:
2022-02-21T02:21:34.9688899Z ==24413==    at 0x4DC1FBB: timerfd_create (syscall-template.S:78)
2022-02-21T02:21:34.9689329Z ==24413==    by 0x4B86DDB: event_setup_timer_fd (sd-event.c:1152)
2022-02-21T02:21:34.9689755Z ==24413==    by 0x4B86F5E: setup_clock_data (sd-event.c:1182)
2022-02-21T02:21:34.9690157Z ==24413==    by 0x4B873DA: sd_event_add_time (sd-event.c:1255)
2022-02-21T02:21:34.9690597Z ==24413==    by 0x4B875B5: sd_event_add_time_relative (sd-event.c:1303)
2022-02-21T02:21:34.9691056Z ==24413==    by 0x110092: test_dhcp6_client (test-dhcp6-client.c:982)
2022-02-21T02:21:34.9691471Z ==24413==    by 0x11047A: main (test-dhcp6-client.c:1018)
2022-02-21T02:21:34.9691733Z ==24413== 
2022-02-21T02:21:34.9691952Z ==24413== Open file descriptor 3:
2022-02-21T02:21:34.9692365Z ==24413==    at 0x4DC1C2B: epoll_create1 (syscall-template.S:78)
2022-02-21T02:21:34.9692770Z ==24413==    by 0x4B841E6: sd_event_new (sd-event.c:399)
2022-02-21T02:21:34.9693211Z ==24413==    by 0x110039: test_dhcp6_client (test-dhcp6-client.c:981)
2022-02-21T02:21:34.9693640Z ==24413==    by 0x11047A: main (test-dhcp6-client.c:1018)
2022-02-21T02:21:34.9693882Z ==24413== 
```

**Pull request merged**